File: backend/routers/compare.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List
from database import get_conn, enrich

router = APIRouter(prefix="/api/compare", tags=["compare"])


class CompareRequest(BaseModel):
    ids: List[int]


FIELDS = [
    ("name",            "College Name"),
    ("city",            "City"),
    ("type",            "Type"),
    ("fees",            "Annual Fees (Rs)"),
    ("placement",       "Placement Rate"),
    ("avg_package",     "Average Package"),
    ("highest_package", "Highest Package"),
    ("seats",           "Total Seats"),
    ("criteria",        "Admission Criteria"),
]


@router.post("/")
def compare_colleges(body: CompareRequest):
    if len(body.ids) < 2 or len(body.ids) > 3:
        raise HTTPException(status_code=400, detail="Provide 2 to 3 college IDs")

    conn = get_conn()
    placeholders = ",".join("?" * len(body.ids))
    rows = conn.execute(
        f"SELECT * FROM colleges WHERE id IN ({placeholders})", body.ids
    ).fetchall()

    if len(rows) != len(body.ids):
        conn.close()
        raise HTTPException(status_code=404, detail="One or more colleges not found")

    colleges = [enrich(conn, dict(r)) for r in rows]
    conn.close()

    return {
        "colleges": colleges,
        "table": [
            {"parameter": label, "values": [c.get(key) for c in colleges]}
            for key, label in FIELDS
        ],
    }

File: backend/routes/compare.js
const express = require('express');
const router = express.Router();
const db = require('../db');

// POST /api/compare - compare 2-3 colleges by IDs
// Body: { ids: [1, 2, 3] }
router.post('/', (req, res) => {
  const { ids } = req.body;

  if (!Array.isArray(ids) || ids.length < 2 || ids.length > 3) {
    return res.status(400).json({ error: 'Provide 2 to 3 college IDs in an array' });
  }

  const placeholders = ids.map(() => '?').join(', ');
  const colleges = db.prepare(`SELECT * FROM colleges WHERE id IN (${placeholders})`).all(...ids);

  if (colleges.length !== ids.length) {
    return res.status(404).json({ error: 'One or more colleges not found' });
  }

  const enriched = colleges.map(college => {
    college.courses = db.prepare('SELECT course FROM college_courses WHERE college_id = ?')
      .all(college.id).map(r => r.course);
    college.branches = db.prepare('SELECT branch FROM college_branches WHERE college_id = ?')
      .all(college.id).map(r => r.branch);
    college.facilities = db.prepare('SELECT facility FROM college_facilities WHERE college_id = ?')
      .all(college.id).map(r => r.facility);
    return college;
  });

  // Build comparison table structure
  const fields = [
    { key: 'name', label: 'College Name' },
    { key: 'city', label: 'City' },
    { key: 'type', label: 'Type' },
    { key: 'fees', label: 'Annual Fees (₹)' },
    { key: 'placement', label: 'Placement Rate' },
    { key: 'avg_package', label: 'Average Package' },
    { key: 'highest_package', label: 'Highest Package' },
    { key: 'seats', label: 'Total Seats' },
    { key: 'criteria', label: 'Admission Criteria' }
  ];

  const comparison = {
    colleges: enriched,
    table: fields.map(field => ({
      parameter: field.label,
      values: enriched.map(c => c[field.key])
    }))
  };

  res.json(comparison);
});

module.exports = router;
